tests/test_basis.py

```diff
@@ -1,203 +1,1024 @@
-
-    Returns
-    -------
-    None
+    An abstract base class that sets the foundation for individual basis function testing.
+    This class requires an implementation of a 'cls' method, which is utilized by the meta-test
+    that verifies if all basis functions are properly tested.
     """
-    for class_name in min_basis_funcs:
-        print(f"\n-> Testing \"{class_name}\"")
-        basis_obj = getattr(basis, class_name)
-        # params that should not raise exception
-        if not min_basis_funcs[class_name]['raise_exception']:
-            basis_obj(**min_basis_funcs[class_name]['args'])
+
+    @abc.abstractmethod
+    def cls(self):
+        pass
+
+
+class TestRaisedCosineLogBasis(BasisFuncsTesting):
+    cls = basis.RaisedCosineBasisLog
+
+    @pytest.mark.parametrize(
+        "args, sample_size",
+        [[{"n_basis_funcs": n_basis}, 100] for n_basis in [2, 10, 100]],
+    )
+    def test_evaluate_returns_expected_number_of_basis(self, args, sample_size):
+        """
+        Verifies the number of basis functions returned by the evaluate() method matches
+        the expected number of basis functions.
+        """
+        basis_obj = self.cls(**args)
+        eval_basis = basis_obj.evaluate(np.linspace(0, 1, sample_size))
+        if eval_basis.shape[0] != args["n_basis_funcs"]:
+            raise ValueError(
+                "Dimensions do not agree: The number of basis should match the first dimension of the evaluated basis."
+                f"The number of basis is {args['n_basis_funcs']}",
+                f"The first dimension of the evaluated basis is {eval_basis.shape[0]}",
+            )
+        return
+
+    @pytest.mark.parametrize("sample_size", [100, 1000])
+    @pytest.mark.parametrize("n_basis_funcs", [2, 10, 100])
+    def test_sample_size_of_evaluate_matches_that_of_input(
+        self, n_basis_funcs, sample_size
+    ):
+        """
+        Checks that the sample size of the output from the evaluate() method matches the input sample size.
+        """
+        basis_obj = self.cls(n_basis_funcs=n_basis_funcs)
+        eval_basis = basis_obj.evaluate(np.linspace(0, 1, sample_size))
+        if eval_basis.shape[1] != sample_size:
+            raise ValueError(
+                f"Dimensions do not agree: The window size should match the second dimension of the evaluated basis."
+                f"The window size is {sample_size}",
+                f"The second dimension of the evaluated basis is {eval_basis.shape[1]}",
+            )
+
+    @pytest.mark.parametrize("n_basis_funcs", [-1, 0, 1, 3, 10, 20])
+    def test_minimum_number_of_basis_required_is_matched(self, n_basis_funcs):
+        """
+        Verifies that the minimum number of basis functions required (i.e., 2) is enforced.
+        """
+        raise_exception = n_basis_funcs < 2
+        if raise_exception:
+            with pytest.raises(ValueError):
```

Comment:
I recommend using `pytest.raises(ValueError, match='contents of error message')` when checking that a specific exceptions is raised (see [here](https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions)), to make sure you got the valueerror you expected